GuessTheNumber.java

```diff
@@ -21,4 +24,129 @@
  * Before the next game, the top 3 scores (which are greater than 0) will be printed
  */
 public class GuessTheNumber {
+
+    private static byte move = 0;
+    private static byte samePosition;
+
+    public static void main(String[] args) {
+
+        Scanner input = new Scanner(System.in);
+        byte[] baseArray = new byte[4];
+        byte[] guessArray = new byte[4];
+        int guessedNumber;
+
+        System.out.printf("Guess the number: start a new game? Yes / No : ");
+        String confirm = input.next();
+
+        if (!confirm.matches("Yes") && (!confirm.matches("yes")) && (!confirm.matches("y"))) {
+            return;
+        }
+
+        System.out.println("A 4-digit number has been chosen. Only digits 0-9 are allowed without duplicates.\n" +
+                "This number is guessed by you via multiple attempts. An attempt consists in a guessed number \n" +
+                "and the computer's reply. The computer must tell you how many digits have you guessed\n" +
+                "on the same position, and how many digits have you guessed on a different position. \n" +
+                "using that reply you must guess the number using as few moves as possible. Good luck! \n");
+
+        baseArray = populateBaseArray(baseArray);
+
+        while (samePosition != 4) {
+            guessedNumber = getUserGuess(input);
+            guessArray = populateGuessArray(guessArray, guessedNumber);
+            samePosition = tryUserGuess(baseArray,guessArray);
+        }
+        System.out.println("You've guessed the number, Good Job!");
+    }
+
+    private static byte[] populateBaseArray(byte[] baseArray) {
+
+        Random randomNum = new Random();
+        byte[] numbersUsed = new byte[10];
+        int i = 0;
+        int nextInt;
+
+        System.out.printf("Computer's num: (");
+
+        while (i < baseArray.length) {
+            nextInt = randomNum.nextInt(10);
+            if (nextInt != 0 && numbersUsed[nextInt] == 0) {
+                baseArray[i] = (byte)nextInt;
+                numbersUsed[nextInt] = 1;
+                System.out.printf(baseArray[i] + " ");
+                i++;
+            }
+        }
+        System.out.printf(")  <== this printout is for debugging purpose only");
+        return baseArray;
+    }
+
+    private static int getUserGuess(Scanner input) {
+        System.out.printf("\nEnter your guess : ");
+        int guessedNumber = input.nextInt();
+        int isNumberValid = checkInputNumberValidity(guessedNumber);
+        while (isNumberValid == 0) {
+            System.out.printf("Enter a valid 4-digits guess: ");
+            guessedNumber = input.nextInt();
+            isNumberValid = checkInputNumberValidity(guessedNumber);
+        }
+        return guessedNumber;

```

Comment:
- remove the isNumberValid variable
- change the return type of checkInputNumberValidity to boolean and return the result (numDigits == 4 && !duplicatesUsed)
- change the name of method from checkInputNumberValidity to isValidGuess to indicate that it returns a boolean
- simplify the loop to
       while (!isValidGuess(guessedNumber)) {
            System.out.printf("Enter a valid 4-digits guess: ");
            guessedNumber = input.nextInt();
        }